tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11400,14 +11400,74 @@ bool hlsl::DiagnoseNodeStructArgument(Sema *self, TemplateArgumentLoc ArgLoc,
   }
 }
 
+bool IsExported(Sema *self, clang::FunctionDecl *FD) {
+  if (FD->hasAttr<HLSLShaderAttr>()) {
+    return true;
+  }
+
+  bool isMarkedStatic = false;
+  bool isMarkedExport = false;
+
+  if (auto cDecl = dyn_cast<CXXMethodDecl>(FD)) {
+    if (cDecl->isStatic()) {
+      isMarkedStatic = true;
+    }
+  }
+
+  if (FD->hasAttr<HLSLExportAttr>()) {
+    isMarkedExport = true;
+  }
+
+  if (isMarkedStatic && isMarkedExport) {
+    self->Diag(FD->getLocation(), diag::err_hlsl_varmodifiersna) << "static"
+                                                                 << "export"
+                                                                 << "function";
+    return false;
+  }
+
+  Linkage L = FD->getLinkageAndVisibility().getLinkage();
+
+  // case 1 requires special assignments to linkage
+  if (L != ExternalLinkage && L != InternalLinkage) {
+    // the condition below implies the shader stage is lib_6_x
+    if (self->getLangOpts().IsHLSLLibrary) {
+      L = ExternalLinkage;
+    } else {
+      L = InternalLinkage;
+    }
+  }
+
+  // now case 2 and 3 can apply and we can determine
+  // whether the function is exported.
+  if (L == InternalLinkage && isMarkedExport) {
+    return true;
+  }
+
+  if (L == ExternalLinkage && isMarkedStatic) {
+    return false;
+  }
+
+  if (L == InternalLinkage) {
+    return false;
+  }
+
+  if (L == ExternalLinkage) {
+    return true;
+  }
+  // unreachable but necessary to prevent warnings
+  return true;
+}
```

Comment:
Talked offline with @tex3D, we decided that it was too complex to use ExportMap by copying it over from DXCOpts to LangOpts (since dxilutil's ExportMap implementation has a memebr that cannot be copied, which is an llvm string map).
We might consider re-addressing this in upstream, but there's also the issue where we don't have mangled names in Sema, and so the export map may not be sufficient information to tell which functions are exported (in case of a name collision).